tls/s2n_handshake.c

```diff
@@ -68,33 +68,70 @@ int s2n_handshake_parse_header(struct s2n_connection *conn, uint8_t * message_ty
     return 0;
 }
 
-int s2n_handshake_get_hash_state(struct s2n_connection *conn, s2n_hash_algorithm hash_alg, struct s2n_hash_state *hash_state)
+static struct s2n_hash_state *s2n_handshake_get_hash_state_ptr(struct s2n_connection *conn, s2n_hash_algorithm hash_alg)
```

Comment:
Why is this change necessary?

Reply:
It is a change in implementation, not in interface (see further down the PR). The comment further down gives a little bit of insight into the differences between EVP and old APIs. 

The function get_hash_state used to return a copy of the hash state. This allowed updating the hash state with new transcript values, but we weren't able to actually reset the hash. When rewriting hash transcripts, we need to reset the hash value.

The comment below describes how you must use `s2n_hash_copy()` to avoid modifying the underlying EVP value. This is pre-existing behavior, not behavior introduced in this change (this is why all the unit tests use `s2n_hash_copy()` already)

Reply:
Ah I see now!

Would it be cleaner to write the signature as "int s2n_handshake_get_hash_state_reference(struct s2n_connection *conn, s2n_hash_algorithm hash_alg, struct s2n_hash_state **hash_state)" and call it like this:
```
    struct s2n_hash_state *hash_state_ptr;
    GUARD(s2n_handshake_get_hash_state_reference(conn, hash_alg, &hash_state_ptr);
```

That way s2n_handshake_get_hash_state_ptr/reference can handle its own errors (callers don't need to know that null result == S2N_ERR_HASH_INVALID_ALGORITHM), and is more in line with how s2n functions usually behave.